diesel_infer_schema/src/codegen.rs

```diff
@@ -0,0 +1,104 @@
+use quote;
+use syn;
+
+use std::error::Error;
+use data_structures::ColumnInformation;
+use inference::{establish_connection, get_table_data, determine_column_type,
+                get_primary_keys, load_table_names, InferConnection};
+
+
+pub fn derive_infer_table_from_schema(database_url: &str, table_name: &str)
+    -> Result<quote::Tokens, Box<Error>>
+{
+
+    let connection = try!(establish_connection(database_url));
+    let data = try!(get_table_data(&connection, table_name));
+    let primary_keys = try!(get_primary_keys(&connection, table_name))
+        .into_iter().map(syn::Ident::new);
+    let table_name = syn::Ident::new(table_name);
+
+    let mut tokens = Vec::with_capacity(data.len());
+    for a in data {
+        tokens.push(try!(column_def_tokens(&a, &connection)));
+    }
+
+    Ok(quote!(table! {
+        #table_name (#(#primary_keys),*) {
+            #(#tokens),*,
+        }
+    }))
+}
+
+pub fn infer_schema_for_schema_name<F1, F2>(
+    database_url: &str,
+    schema_name: Option<&str>,
+    error_handler: F1,
```

Comment:
Taking an error handler function seems unideomatic compared to returning an error.